File: scripts/encrypt_mt103.py
from cryptography.fernet import Fernet
import base64
import json

class MT103Encryptor:
    def __init__(self):
        self.key = Fernet.generate_key()
        self.cipher_suite = Fernet(self.key)

    def encrypt_message(self, mt103_data):
        """
        Encrypts MT103 message data using Fernet (symmetric encryption)
        """
        if isinstance(mt103_data, dict):
            mt103_data = json.dumps(mt103_data)
        
        encrypted_data = self.cipher_suite.encrypt(mt103_data.encode())
        return base64.urlsafe_b64encode(encrypted_data)

    def decrypt_message(self, encrypted_data):
        """
        Decrypts MT103 message data
        """
        try:
            decrypted_data = self.cipher_suite.decrypt(base64.urlsafe_b64decode(encrypted_data))
            return json.loads(decrypted_data)
        except Exception as e:
            return {"error": f"Decryption failed: {str(e)}"}

if __name__ == "__main__":
    # Example usage
    encryptor = MT103Encryptor()
    test_data = {
        "messageType": "MT103",
        "sender": "BANKXXXX",
        "receiver": "BANKYYYY",
        "amount": "1000.00",
        "currency": "USD"
    }
    
    encrypted = encryptor.encrypt_message(test_data)
    print(f"Encrypted: {encrypted}")
    
    decrypted = encryptor.decrypt_message(encrypted)
    print(f"Decrypted: {decrypted}")

File: scripts/encrypt_mt103.js
const crypto = require('crypto');

class MT103Encryptor {
    constructor() {
        this.algorithm = 'aes-256-gcm';
        this.key = crypto.randomBytes(32);
    }

    encryptMessage(mt103Data) {
        try {
            const iv = crypto.randomBytes(16);
            const cipher = crypto.createCipheriv(this.algorithm, this.key, iv);
            
            const data = typeof mt103Data === 'string' ? 
                mt103Data : JSON.stringify(mt103Data);
            
            let encrypted = cipher.update(data, 'utf8', 'base64');
            encrypted += cipher.final('base64');
            
            return {
                encrypted,
                iv: iv.toString('base64'),
                authTag: cipher.getAuthTag().toString('base64')
            };
        } catch (error) {
            console.error('Encryption failed:', error);
            throw new Error('Encryption failed');
        }
    }

    decryptMessage(encryptedData) {
        try {
            const decipher = crypto.createDecipheriv(
                this.algorithm, 
                this.key,
                Buffer.from(encryptedData.iv, 'base64')
            );
            
            decipher.setAuthTag(Buffer.from(encryptedData.authTag, 'base64'));
            
            let decrypted = decipher.update(encryptedData.encrypted, 'base64', 'utf8');
            decrypted += decipher.final('utf8');
            
            return JSON.parse(decrypted);
        } catch (error) {
            return { error: `Decryption failed: ${error.message}` };
        }
    }
}

module.exports = new MT103Encryptor();
